### PR TITLE
Refactor implementation of `stack_snapshot`

### PR DESCRIPTION
**Pull request by aherrmann**
This PR extracts the refactorings of the `stack_snapshot` implementation that occurred in https://github.com/tweag/rules_haskell/issues/1310. It does not change the functionality of `stack_snapshot` at this point. I have confirmed that the results of `bazel fetch` don't change between `master` and this PR on `@stackage`, `@stackage-zlib`, `@ghcide`, and `@stackge` in `examples`.

- Parse instead of validating the output of `stack ls dependencies json`. In particular, consistently introduce a `location.type` field into a package specification. Before, core packages had no `location` field at all.
- Factor out checks on the `snapshot`/`local_snapshot` attributes to make them re-usable between pinned and non-pinned `stack_snapshot`.
- Factor out parsing of the user provided package list to make them re-usable between pinned and non-pinned `stack_snapshot`.
- Split `_compute_dependency_graph` into two separate functions `_resolve_packages` and `_download_packages`. The pinned version will only need to call a version of `_download_packages` but not `_resolve_packages`. This also simplifies the transformations on the output of `stack ls dependencies json` that `_compute_dependency_graph` used to perform for historical reasons. The output of `stack ls dependencies json` is already pretty close to what we need in `stack_snapshot`. The new format is also easier to de-/serialize through JSON. Notably, `_resolve_packages` sorts the packages by name to ensure that the generated `BUILD.bazel` file and `packages.bzl` are reproducible.
- The generated `packages.bzl` file is modified to contain one package per line instead of all packages in one line. This is to make the file easier to read for humans and friendlier towards diffing.

### CODE REVIEW

File: haskell/cabal.bzl [hunk 1]
```diff
@@ -892,23 +892,71 @@ def _get_components(components, package):
...
+    actual_ty = type(json[field])
+    if actual_ty != ty:
+        fail(fmt % "Expected field '{field}' of type '{expected}', but got '{got}'.".format(
+            field = field,
+            expected = ty,
+            got = actual_ty,
+        ))
+    return json[field]
+
+def _parse_package_spec(package_spec):
```
Comment (Profpatsch) [inline haskell/cabal.bzl:920]: A link to the definition of the package spec format would be nice (though I don’t have the illusion that there exists one)

File: haskell/cabal.bzl [hunk 2]
```diff
@@ -892,23 +892,71 @@ def _get_components(components, package):
...
+        field: _parse_json_field(package_spec, field, ty, errmsg.format(context = "package description"))
+        for (field, ty) in [("name", "string"), ("version", "string"), ("dependencies", "list")]
+    }
+
+    # Parse location field.
+    location = {}
+    if parsed["name"] in _CORE_PACKAGES or not "location" in package_spec:
+        location["type"] = "core"
+    else:
+        location_type = _parse_json_field(package_spec["location"], "type", "string", errmsg.format(context = "location description"))
```
Comment (Profpatsch) [inline haskell/cabal.bzl:934]: Some really long lines from here

File: haskell/cabal.bzl [hunk 3]
```diff
@@ -892,23 +892,71 @@ def _get_components(components, package):
...
+        if location_type == "project package":
+            location["type"] = "vendored"
+        elif location_type == "hackage":
+            location["type"] = location_type
+            url_prefix = _parse_json_field(package_spec["location"], "url", "string", errmsg.format(context = location_type + " location description"))
+            location["url"] = url_prefix + "/{name}-{version}.tar.gz".format(**parsed)
+            # stack does not expose sha-256, see https://github.com/commercialhaskell/stack/issues/5274
```
Comment (Profpatsch) [inline haskell/cabal.bzl:941]: why is that important?

**Comment by Profpatsch** [comment]
I didn’t review the second-to-last commit “Separate resolving and downloading packages”, as it was kind of hard to follow which parts went where, but I guess we’ll see if the changes are correct. It would probably be easier to review if it separated the parts that remove the legacy transformations from the pure refactor code.

File: haskell/cabal.bzl (hunk 1 above)
Comment (aherrmann) [reply haskell/cabal.bzl:920]: That's a good idea. I'm not aware of a standalone definition, however the `stack` sources certainly define the current format. I've added a link to the docstring.

File: haskell/cabal.bzl (hunk 2 above)
Comment (aherrmann) [reply haskell/cabal.bzl:934]: Fixed.

File: haskell/cabal.bzl (hunk 3 above)
Comment (aherrmann) [reply haskell/cabal.bzl:941]: Once we download Hackage and archive dependencies using Bazel rather than `stack unpack` we'll want the SHA-256 to make sure the download is reproducible and caches well in Bazel's repository cache. However, since `stack ls dependencies json` doesn't expose the SHA-256 we have to determine it in other ways. https://github.com/tweag/rules_haskell/pull/1376 implements that.

If `stack` implements https://github.com/commercialhaskell/stack/issues/5274 then we can directly extract the SHA-256 here instead of determining it ourselves later on.
